### DISCUSSION TITLE
tutorials/voyage-part2/10-implementing-resolvers

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#152] [course feedback]
# tutorials/voyage-part2/10-implementing-resolvers

https://www.apollographql.com/tutorials/voyage-part2/10-implementing-resolvers

**Comment by shayk89** [comment]
It's confusing that the 'hasPictures' property we're told to use in the exercise isn't present in the **schema** we're told to use

**Reply by lizhennessy** [reply]
Thanks for pointing this out, @shayk89! I'll add a clarification that `hasPictures` is available on the book _object_, but not necessarily defined in the schema. (Though we certainly _could_ add it if need be! :) )

**Comment by chrisspinner** [comment]
I am still not clear on why "role" is passed as the third item in the return object for author. I am confused with the reasoning above. I thought this is the reason we set __typename to role.

**Reply by lizhennessy** [reply]
Thanks for bringing this to our attention @chrisspinner ! You're absolutely right — in this example, setting `__typename: role` explicitly is actually overridden once the object arrives in the `User.resolveType` resolver (receiving instead a `__typename` of `User`). We're updating the course now; your `Review.author` resolver will work just fine returning the following:

```
Review: {
  author: (review) => {
    let role = '';
    if (review.targetType === 'LISTING' || review.targetType === 'HOST') {
      role = 'Guest';
    } else {
      role = 'Host';
    }
    return { id: review.authorId, role };
  },
},
```

Great catch, thank you! Please let us know if you encounter anything else that could use further clarification or attention.

**Comment by ibrahimnadra** [comment]
how is the first param : user? As i know the first one is the entity representation which is composed of typename and key, that is id 
Guest: {
  __resolveReference: (user, { dataSources }) => {
    return dataSources.accountsAPI.getUser(user.id);
  },
},

**Comment by ibrahimnadra** [comment]
Why do we add the typename here? As in the previous section, i read: "We don't need to include the __typename field in this entity representation. The router automatically knows how to get the __typename field (because __typename is a meta-field that's included on all objects, per the GraphQL specification)"
.const type = book.hasPictures ? 'PictureBook' : 'YoungAdultNovel';
return {__typename: type, isbn: book.isbn}
